### PR TITLE
First draft of the function allow to run the simulation (run_sim)

### PR DESCRIPTION
**Pull request by francoiseblanc**
Added the run_sim function.

### CODE REVIEW

**Review by AidanFahlman — APPROVED** [review]
This all looks great Francoise, couldn't find anything to change or tweak. Great job!

File: src/simulation_lifecycle/simulation.c [hunk 1]
```diff
@@ -32,7 +35,7 @@ int parse_default_sim_config(const cJSON *simulation_config, cJSON *target);
  */
 int write_sim_config(const cJSON *simulation_config, char *config_json);
 
-int build_simulation_scenario(cJSON *simulation_config, Node_t  **data_sources) {
+int build_simulation_scenario(cJSON *simulation_config, feature_set_t *p_features, relation_set_t *p_relations) {
```
Comment (romancardenas) [inline src/simulation_lifecycle/simulation.c:38]: This function definition is outdated. Please, change it to 
`int build_simulation_scenario(cJSON *simulation_config, node_t  **data_sources) {`

File: src/simulation_lifecycle/simulation.c [hunk 2]
```diff
@@ -89,3 +92,58 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
 }
+
+int run_sim(const cJSON *simulation_config){
+
+    /* Get model name - e.g. 2_2_agent_sir_config */
+    cJSON *model_id = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_MODEL_ID);
+    char *model = cJSON_GetStringValue(model_id);
+
+    /* Define executable - e.g. 2_2_agent_sir_config.exe */
+    char *model_executable = concat(model, ".exe");
```
Comment (romancardenas) [inline src/simulation_lifecycle/simulation.c:103]: This is only for Windows. In MacOS, for instance, the binary has no extension at all.

We must add a macro definition, something like:
```
#ifdef WIN32
    model = concat(model, ".exe");
#endif
```

File: src/simulation_lifecycle/simulation.c [hunk 3]
```diff
@@ -89,3 +92,58 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
+    char *config_path = cJSON_GetStringValue(config);
+
+    /* Define a command and run simulation:
+     * The command is changing to the directory containing the .exe file.
+     * Then, the command executes the .exe file with one argument
+     * which is the scenario config json file.
+     * Example: cd ../third_party/CellDEVS_models/tutorial/bin/ ;
+     * ./2_2_agent_sir_config.exe ../../../../test/data/run_simulation/config.json
+     */
+    char *command = concat("cd ",SIM_MODEL_LIBRARY);
```
Comment (romancardenas) [inline src/simulation_lifecycle/simulation.c:116]: look at `sprintf` function. It is more readable and user friendly than doing multiple concatenations ;)
https://stackoverflow.com/questions/804288/creating-c-formatted-strings-not-printing-them

File: src/simulation_lifecycle/simulation.c [hunk 4]
```diff
@@ -89,3 +92,58 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
+        char *results_filename = concat(model, SIM_RESULTS_END_FILENAME);
+        char *results_filename_path = concat(SIM_RESULTS_DEFAULT_PATH,results_filename);
+
+        /* Check if results file exists */
+        if (!file_exists(results_filename_path)) {
+            return SIM_RUN_NO_RESULTS;
+        }else{
+            /* Moving the results file to predefined result output path. */
+            cJSON *result = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_RESULT_OUTPUT_PATH);
+            char *result_path = cJSON_GetStringValue(result);
```
Comment (romancardenas) [inline src/simulation_lifecycle/simulation.c:137]: You must ensure that `result_path` is not `NULL`. I usually do:
```
char *result_path = cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_RESULT_OUTPUT_PATH);
if (result_path == NULL) {
    return CORRESPONDING_ERROR;
}
...
```

File: data/simulation_config.json [hunk 5]
```diff
@@ -34,6 +34,6 @@
       }
     }
   },
-  "config_output_path": "./config_output.json",
-  "result_output_path": "./result_output.json"
+  "config_output_path": "../data/simulation_config.json",
+  "result_output_path": "..data/results/"
```
Comment (romancardenas) [inline data/simulation_config.json:38]: If it is going to be a directory, I'd call it `results_output_dir`.
Additionally, we have to be careful: maybe the user didn't add the last `/`, this is a common issue when dealing with directories.

**Review by romancardenas — CHANGES_REQUESTED** [review]
Good job, @francoiseblanc ! I added some important comments, please, address them and I'll review your function again.

**Comment by romancardenas** [comment]
For the windows issue (the `.exe` extension), we can add these lines to the `CMakeList.txt`file:

```
if (WIN32)
    add_compile_definitions(WITH_WINDOWS)
endif()
```

Doing so, `cmake` will define the `WITH_WINDOWS` macro if and only if the machine that is compiling the project is a Windows machine. Thus, your `concat` function to add the `.exe` extension will run only on Windows machines.

Give it a try!

File: src/simulation_lifecycle/simulation.c (hunk 2 above)
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:103]: In the Cygwin terminal or the Windows Command Prompt, inserting ".exe" or not works. So I can just remove the extension. However, the Windows Command Prompt, won't take "./". Do we want to be able to use the Windows Command Prompt or just the Cygwin terminal?

File: src/simulation_lifecycle/simulation.c (hunk 4 above)
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:137]: I did not do these checks since this function is always executed after other functions that make sure that the json file contains all the info. I guess there is no harm in adding these extra checks :)

Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:137]: Actually, do we want to redo these checks? I would have to do it also for:
`cJSON *model_id = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_MODEL_ID);`
`cJSON *config = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_CONFIG_OUTPUT_PATH);`

File: data/simulation_config.json (hunk 5 above)
Comment (francoiseblanc) [reply data/simulation_config.json:38]: I agree that using directories may be not ideal. If we want to use path, I think we need two result output paths as follows:
"result_output_json_path": "../data/results/simulation_results.json
"result_output_messages_path": "../data/results/simulation_results.json

Could you confirm which .json file would have to be used for the conversion? Is it actually "config_output_path"? If it is, I'll just do something like that:

"config_output_path": "../data/results/simulation_config.json
"result_output_path": "../data/results/simulation_results.txt

File: src/simulation_lifecycle/simulation.c (hunk 2 above)
Comment (romancardenas) [reply src/simulation_lifecycle/simulation.c:103]: How does it work in the Windows command prompt?

If it is easy to adapt, we can try

File: src/simulation_lifecycle/simulation.c (hunk 4 above)
Comment (romancardenas) [reply src/simulation_lifecycle/simulation.c:137]: `I think that each function should check their corresponding parts (mine does so).
For example, the `validate_workflow` won't check the complete validity of the workflow, but it checks that at first sight, it looks OK (we can discuss this on Friday).

File: data/simulation_config.json (hunk 5 above)
Comment (romancardenas) [reply data/simulation_config.json:38]: Good question... Maybe @staubibr can help

Comment (romancardenas) [reply data/simulation_config.json:38]: The config output path is used by Cadmium to run the simulation.
The result output path is where the result of Cadmium should be placed.
I think that the viewer only uses the output messages, not the state

Comment (francoiseblanc) [reply data/simulation_config.json:38]: The viewer uses a .json file and the output messages (not the states). @staubibr Which .json file is used for the viewer?

File: src/simulation_lifecycle/simulation.c (hunk 4 above)
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:137]: Do you mean that for model_id, for example, I should verify if empty, null, does not exist, etc. and do the appropriate tests?

File: src/simulation_lifecycle/simulation.c (hunk 2 above)
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:103]: Should be easy to adapt. I'll look into it.

File: CMakeLists.txt [hunk 6]
```diff
@@ -6,7 +6,8 @@ set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/build)
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/build)
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/bin")
 project(${PROJECT_NAME})
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c11")
+#set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c11")
+set(CMAKE_C_STANDARD 11)
```
Comment (francoiseblanc) [inline CMakeLists.txt:10]: `set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c11")` would not change my version of C. It would remain to C99.

Comment (francoiseblanc) [reply CMakeLists.txt:10]: @romancardenas I would appreciate if you could check if this change is okay.

File: src/simulation_lifecycle/simulation.c (hunk 4 above)
Comment (romancardenas) [reply src/simulation_lifecycle/simulation.c:179]: Not that part, because I already checked in the `build_simulation_scenario`. But the path to the results yes (this function is the very first one that reads it)

File: src/simulation_lifecycle/simulation.c [hunk 7]
```diff
@@ -69,7 +73,7 @@ int parse_default_sim_config(const cJSON *simulation_config, cJSON *target) {
     }
     // Check that the model exists
     char *model_path = concat(SIM_MODEL_LIBRARY, cJSON_GetStringValue(model));
-    int model_found = executable_exists(model_path);
+    int model_found = file_exists(model_path);
```
Comment (romancardenas) [inline src/simulation_lifecycle/simulation.c:76]: Why did you change this? The executable file needs execution permissions, not read permissions. Check the difference between the functions `executable_exists` and `file_exists`

File: src/simulation_lifecycle/simulation.c [hunk 8]
```diff
@@ -89,3 +93,103 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
+    cJSON *model = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_MODEL_ID);
+
+    /* Check that model ID is provided using a valid format. */
+    if (model == NULL || !cJSON_IsString(model)) {
+        return SIM_MODEL_SELECTION_INVALID;
+    }
+
+    /* Check that the model exists. */
+    char *model_path = concat(SIM_MODEL_LIBRARY, cJSON_GetStringValue(model));
+    int model_found = file_exists(model_path);
```
Comment (romancardenas) [inline src/simulation_lifecycle/simulation.c:109]: Same here, you should use the `executable_exists` function

File: CMakeLists.txt (hunk 6 above)
Comment (romancardenas) [reply CMakeLists.txt:10]: It seems to be a new functionality in CMake 3.1. I'm OK with the change

**Review by romancardenas — APPROVED** [review]
Good work, @francoiseblanc. I left some comments, can you address them before merging?

File: src/simulation_lifecycle/simulation.c (hunk 7 above)
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:76]: It would not work but I'll check again the difference between the functions.

Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:76]: I've made the change at both places. It is working as expected.

File: src/simulation_lifecycle/simulation.c (hunk 8 above)
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:109]: Change made.

File: src/simulation_lifecycle/simulation.c [hunk 9]
```diff
@@ -89,3 +93,103 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
+    getcwd(current_dir, sizeof(current_dir));
+
+    /* Change to model executable files directory. */
+    chdir(SIM_MODEL_LIBRARY);
+
+    /* Get config output path which is relative to SIM_MODEL_LIBRARY. */
+    cJSON *config = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_CONFIG_OUTPUT_PATH);
+
+    /* Check that config output path is provided using a valid format. */
+    if (config == NULL || !cJSON_IsString(config)) {
```
[the remark this answers is not in the document]
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:127]: Great recommendation!!! I've made the change and also did it for `result_path`. It cleans up the code pretty well.

File: src/simulation_lifecycle/simulation.c [hunk 10]
```diff
@@ -89,3 +93,103 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
+        char *results_filename = concat(model_name, SIM_RESULTS_END_FILENAME);
+        char *results_filename_path = concat(SIM_RESULTS_DEFAULT_PATH,results_filename);
+
+        /* Check if results file exists. */
+        if (!file_exists(results_filename_path)) {
+            return SIM_RUN_NO_RESULTS;
+        }else{
+            /* Moving the results file to predefined result output path. */
+            /* Get result output path */
+            cJSON *result = cJSON_GetObjectItemCaseSensitive(simulation_config, SIM_RESULT_OUTPUT_PATH);
```
[the remark this answers is not in the document]
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:170]: Done.

File: src/simulation_lifecycle/simulation.c [hunk 11]
```diff
@@ -89,3 +93,103 @@ int write_sim_config(const cJSON *simulation_config, char *config_json) {
...
+     * The command executes the model executable file with one argument
+     * which is the scenario config .json file.
+     * Example: ./2_2_agent_sir_config ../../../../test/data/run_simulation/config.json
+     */
+
+    /* TODO Issue: "./" works on Linux. On Windows, it will work with Windows PowerShell, but not CMD */
+    /* TODO Find a better solution than using a relative config output path from "../third_party/CellDEVS_models/tutorial/bin/" */
+
+    char command[MAX_LEN] = "";
+    snprintf(command,sizeof(command),"./%s %s",model_name,config_path);
```
[the remark this answers is not in the document]
Comment (francoiseblanc) [reply src/simulation_lifecycle/simulation.c:142]: Agreed.

**Review by romancardenas — APPROVED** [review]
Nice!
